File: src/Calculator.Shared/Common/NetworkManager.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Text;
using Windows.Networking.Connectivity;

namespace CalculatorApp
{
	public enum NetworkAccessBehavior
    {
        Normal = 0,
        OptIn = 1,
        Offline = 2
    };

	public delegate void NetworkBehaviorChangedHandler(NetworkAccessBehavior behavior);

	public sealed class NetworkManager
	{
		public event NetworkBehaviorChangedHandler NetworkBehaviorChanged;

		public NetworkManager()
		{
			// TODO: UNO (Not implemented)
			//NetworkInformation.NetworkStatusChanged += new NetworkStatusChangedEventHandler(OnNetworkStatusChange);
		}

		~NetworkManager()
		{
			// TODO: UNO (Not implemented)
			//NetworkInformation.NetworkStatusChanged -= OnNetworkStatusChange;
		}

		public static NetworkAccessBehavior GetNetworkAccessBehavior()
		{
			// TODO: UNO (Not implemented)
			return NetworkAccessBehavior.Normal;

			//NetworkAccessBehavior behavior = NetworkAccessBehavior.Offline;
			//ConnectionProfile connectionProfile = NetworkInformation.GetInternetConnectionProfile();
			//if (connectionProfile != null)
			//{
			//	NetworkConnectivityLevel connectivityLevel = connectionProfile.GetNetworkConnectivityLevel();
			//	if (connectivityLevel == NetworkConnectivityLevel.InternetAccess || connectivityLevel == NetworkConnectivityLevel.ConstrainedInternetAccess)
			//	{
			//		ConnectionCost connectionCost = connectionProfile.GetConnectionCost();
			//		behavior = ConvertCostInfoToBehavior(connectionCost);
			//	}
			//}

			//return behavior;
		}

		void OnNetworkStatusChange(object sender)
		{
			NetworkBehaviorChanged?.Invoke(GetNetworkAccessBehavior());
		}

		// See app behavior guidelines at https://msdn.microsoft.com/en-us/library/windows/apps/xaml/jj835821(v=win.10).aspx
		static NetworkAccessBehavior ConvertCostInfoToBehavior(ConnectionCost connectionCost)
		{
			if (connectionCost.Roaming || connectionCost.OverDataLimit || connectionCost.NetworkCostType == NetworkCostType.Variable
				|| connectionCost.NetworkCostType == NetworkCostType.Fixed)
			{
				return NetworkAccessBehavior.OptIn;
			}

			return NetworkAccessBehavior.Normal;
		}
	}
}

File: src/CalcViewModel/Common/NetworkManager.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "NetworkManager.h"

using namespace CalculatorApp;
using namespace CalculatorApp::ViewModel;
using namespace CalculatorApp::ViewModel::Common;
using namespace Platform;
using namespace Windows::Networking::Connectivity;

NetworkManager::NetworkManager()
{
    m_NetworkStatusChangedToken = NetworkInformation::NetworkStatusChanged +=
        ref new NetworkStatusChangedEventHandler(this, &NetworkManager::OnNetworkStatusChange, CallbackContext::Same);
}

NetworkManager::~NetworkManager()
{
    NetworkInformation::NetworkStatusChanged -= m_NetworkStatusChangedToken;
}

NetworkAccessBehavior NetworkManager::GetNetworkAccessBehavior()
{
    NetworkAccessBehavior behavior = NetworkAccessBehavior::Offline;
    ConnectionProfile ^ connectionProfile = NetworkInformation::GetInternetConnectionProfile();
    if (connectionProfile != nullptr)
    {
        NetworkConnectivityLevel connectivityLevel = connectionProfile->GetNetworkConnectivityLevel();
        if (connectivityLevel == NetworkConnectivityLevel::InternetAccess || connectivityLevel == NetworkConnectivityLevel::ConstrainedInternetAccess)
        {
            ConnectionCost ^ connectionCost = connectionProfile->GetConnectionCost();
            behavior = ConvertCostInfoToBehavior(connectionCost);
        }
    }

    return behavior;
}

void NetworkManager::OnNetworkStatusChange(_In_ Object ^ /*sender*/)
{
    NetworkBehaviorChanged(GetNetworkAccessBehavior());
}

// See app behavior guidelines at https://msdn.microsoft.com/en-us/library/windows/apps/xaml/jj835821(v=win.10).aspx
NetworkAccessBehavior NetworkManager::ConvertCostInfoToBehavior(_In_ ConnectionCost ^ connectionCost)
{
    if (connectionCost->Roaming || connectionCost->OverDataLimit || connectionCost->NetworkCostType == NetworkCostType::Variable
        || connectionCost->NetworkCostType == NetworkCostType::Fixed)
    {
        return NetworkAccessBehavior::OptIn;
    }

    return NetworkAccessBehavior::Normal;
}
